### PR TITLE
Expose load options capi

### PR DESCRIPTION
**Pull request by sg20180546**
check my code please #9233 
@ajkr

### CODE REVIEW

File: db/c.cc [hunk 1]
```diff
@@ -167,6 +171,7 @@ struct rocksdb_cache_t {
 };
 struct rocksdb_livefiles_t       { std::vector<LiveFileMetaData> rep; };
 struct rocksdb_column_family_handle_t  { ColumnFamilyHandle* rep; };
+struct rocksdb_column_family_descriptor_t {std::vector<ColumnFamilyDescriptor> rep;};
```
Comment (ajkr) [inline db/c.cc:174]: s/descriptor/descriptors/

File: db/c.cc [hunk 2]
```diff
@@ -205,6 +210,11 @@ struct rocksdb_optimistictransaction_options_t {
 struct rocksdb_compactionfiltercontext_t {
   CompactionFilter::Context rep;
 };
+struct rocksdb_loaded_options_t{
+  std::vector<ColumnFamilyDescriptor> cf;
+  DBOptions* dboption;
+};
```
Comment (ajkr) [inline db/c.cc:216]: We don't need this struct, do we? Can the new functions return void?

**Review by ajkr — CHANGES_REQUESTED** [review]
Mostly looks fine with me.

**Comment by sg20180546** [comment]
@ajkr I change code with your feedback. Thank you!

**Comment by ajkr** [comment]
Thanks for the updates. Do you have a test or explanation how the user can open a DB using options loaded from a file? This is the relevant DB open function I found: https://github.com/facebook/rocksdb/blob/d7c868b0624518134147ee98877c0f41d531c56f/db/c.cc#L825-L829

**Comment by sg20180546** [comment]
@ajkr I write test code, run test, and passed in my computer.
but one thinng that on my mind is options_filename in the c_test (snprintf).
thank you!

**Review by ajkr — COMMENTED** [review]
Thanks for the tests. I do not understand how `rocksdb_dboptions_t` and `rocksdb_column_family_descriptors_t` can be used after `rocksdb_load_latest_options()` populates them. For example, I do not see how we can pass them to a function like `rocksdb_open()`.

**Comment by sg20180546** [comment]
`  // Load the options file.
  DBOptions loaded_db_opt;
  std::vector<ColumnFamilyDescriptor> loaded_cf_descs;
  ConfigOptions config_options;
  s = LoadLatestOptions(config_options, kDBPath, &loaded_db_opt,
                        &loaded_cf_descs);
  assert(s.ok());
  assert(loaded_db_opt.create_if_missing == db_opt.create_if_missing);

  // Initialize pointer options for each column family
  for (size_t i = 0; i < loaded_cf_descs.size(); ++i) {
    auto* loaded_bbt_opt =
        loaded_cf_descs[0]
            .options.table_factory->GetOptions<BlockBasedTableOptions>();
    // Expect the same as BlockBasedTableOptions will be loaded form file.
    assert(loaded_bbt_opt->block_size == bbt_opts.block_size);
    // However, block_cache needs to be manually initialized as documented
    // in rocksdb/utilities/options_util.h.
    loaded_bbt_opt->block_cache = cache;
  }
  // In addition, as pointer options are initialized with default value,
  // we need to properly initialized all the pointer options if non-defalut
  // values are used before calling DB::Open().
  assert(loaded_cf_descs[0].options.compaction_filter == nullptr);
  loaded_cf_descs[0].options.compaction_filter = compaction_filter.get();

  // reopen the db using the loaded options.
  std::vector<ColumnFamilyHandle*> handles;
  s = DB::Open(loaded_db_opt, kDBPath, loaded_cf_descs, &handles, &db);
  assert(s.ok());
`

**Comment by sg20180546** [comment]
`  // Load the options file.
  DBOptions loaded_db_opt;
  std::vector<ColumnFamilyDescriptor> loaded_cf_descs;
  ConfigOptions config_options;
  s = LoadLatestOptions(config_options, kDBPath, &loaded_db_opt,
                        &loaded_cf_descs);
  assert(s.ok());
  assert(loaded_db_opt.create_if_missing == db_opt.create_if_missing);

  // Initialize pointer options for each column family
  for (size_t i = 0; i < loaded_cf_descs.size(); ++i) {
    auto* loaded_bbt_opt =
        loaded_cf_descs[0]
            .options.table_factory->GetOptions<BlockBasedTableOptions>();
    // Expect the same as BlockBasedTableOptions will be loaded form file.
    assert(loaded_bbt_opt->block_size == bbt_opts.block_size);
    // However, block_cache needs to be manually initialized as documented
    // in rocksdb/utilities/options_util.h.
    loaded_bbt_opt->block_cache = cache;
  }
  // In addition, as pointer options are initialized with default value,
  // we need to properly initialized all the pointer options if non-defalut
  // values are used before calling DB::Open().
  assert(loaded_cf_descs[0].options.compaction_filter == nullptr);
  loaded_cf_descs[0].options.compaction_filter = compaction_filter.get();

  // reopen the db using the loaded options.
  std::vector<ColumnFamilyHandle*> handles;
  s = DB::Open(loaded_db_opt, kDBPath, loaded_cf_descs, &handles, &db);
  assert(s.ok());
`

**Comment by sg20180546** [comment]
@riversand963

**Review by mrambacher — CHANGES_REQUESTED** [review]
I think this is more complicated than it needs to be.  I suggest only adding three new APIs:

extern ROCKSDB_LIBRARY_API int rocksdb_load_options_from_file(const char* filename,
rocksdb_env_t* env, int ignore_unknown_optons,
rocksdb_cache_t* cache, 
rocksdb_options_t*** options,
char*** column_families,
char** errptr);

extern ROCKSDB_LIBRARY_API int rocksdb_load_latest_options(const char* dbpath,
rocksdb_env_t* env, int ignore_unknown_optons,
rocksdb_cache_t* cache, 
rocksdb_options_t*** options,
char*** column_families,
char** errptr);

extern ROCKSDB_LIBRARY_API void rocksdb_list_options_destroy(rocksdb_options_t** list, int sz);

The "load" methods will give you a list of options and column family names suitable for calling the DB create methods for column families.  The return code can be the size of the two lists (options_t, column_family_names).  

Note that there is already a method for deleting a list of column family names, so a new one is not needed.

**Comment by sg20180546** [comment]
x

**Pull request closed without merging**